src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Order/TreeItemOrderPropertyProvider.cs

```diff
@@ -0,0 +1,108 @@
+            // are mapped in in _rootedOrderMap only; everything else is mapped in
+            // _displayOrderMap only.
+            foreach (var item in OrderedItems)
+            {
+                var includeParts = item.EvaluatedInclude.Split(CommonConstants.PathSeparators, StringSplitOptions.RemoveEmptyEntries);
+                foreach (var part in includeParts)
+                {
+                    var rootedPath = duplicateFiles.Contains(part) ? _project.MakeRooted(item.EvaluatedInclude) : null;
+                    if (rootedPath != null && !_rootedOrderMap.ContainsKey(rootedPath))
+                    {
+                        _rootedOrderMap.Add(rootedPath, index++);
+                    }
+                    else if (!_displayOrderMap.ContainsKey(part))
+                    {
+                        _displayOrderMap.Add(part, index++);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Assign a display order property to items that have previously been preordered
+        /// or other (hidden) items under the project root that are not folders
+        /// </summary>
+        /// <param name="propertyContext">context for the tree item being evaluated</param>
+        /// <param name="propertyValues">mutable properties that can be changed to affect display order etc</param>
+        public void CalculatePropertyValues(
+            IProjectTreeCustomizablePropertyContext propertyContext, 
+            IProjectTreeCustomizablePropertyValues propertyValues)
+        {
+            if (propertyValues is IProjectTreeCustomizablePropertyValues2 propertyValues2)
+            {
+                var isAllowedItemType = propertyContext.ItemType != null 
+                    && _allowedItemTypes.Contains(propertyContext.ItemType);
+
+                // assign display order to folders and items that have a recognized 
+                // item type and appear in order map
+                bool hasDisplayOrder = false;
+                if (isAllowedItemType || propertyContext.IsFolder)
+                {
+                    if (_displayOrderMap.TryGetValue(propertyContext.ItemName, out var index) 
+                        || (propertyContext.Metadata.TryGetValue(FullPathProperty, out var fullPath)
+                            && _rootedOrderMap.TryGetValue(fullPath, out index)))
+                    {
+                        propertyValues2.DisplayOrder = index;
+                        hasDisplayOrder = true;
+                    }
+                }
+
+                if (!hasDisplayOrder && propertyContext.ParentNodeFlags.Contains(ProjectTreeFlags.ProjectRoot) && !propertyContext.IsFolder)
+                {
+                    // move unordered non-folder items at project root to the end 
+                    // (this will typically be hidden items visible on "Show All Files")
+                    propertyValues2.DisplayOrder = _displayOrderMap.Count + 1;
```

Comment:
I think you setset them to Int.Max, or something which doesn't change.  That will reduce the overhead to shuffling them.

Reply:
Also, should you move hidden items in other folders to the end as well?  Do you really need check the parent is the project root?